spec/rails_integration_spec.rb

```diff
@@ -106,14 +106,12 @@ def prepare_params(params)
   end
 
   describe "nested_array" do
-    it "responds with a 400 when the nested array is not supplied properly" do
```

Comment:
What's the value here in removing that test case tho? If the description of the test was inaccurate, let's address that. However, this was an edge case bug that someone encountered and added the specs for, so let's keep that coverage.

Reply:
@ifellinaholeonce fair do's. I have assumed that someone just did wrote a rather silly test. I have restored it and it is now failing.

Reply:
The `ParamEvaluator` is now receiving the following params:

```
#<ActionController::Parameters {"filter"=>"state", "controller"=>"fake", "action"=>"nested_array"} permitted: false>
{"state"=>nil}
```

Reply:
From my understanding `HashParam` will happily coerce `filter: "state"` into `filter: { state: nil }` which will result in the glorious:

```
       expected RailsParam::InvalidParameterError, got #<NoMethodError: undefined method `each_with_index' for nil:NilClass> with backtrace:
         # ./lib/rails_param/param_evaluator.rb:51:in `recurse_on_parameter'
         # ./lib/rails_param/param_evaluator.rb:35:in `param!'
         # ./spec/fixtures/controllers.rb:37:in `block in nested_array'
         # ./lib/rails_param/param_evaluator.rb:66:in `recurse'
         # ./lib/rails_param/param_evaluator.rb:59:in `recurse_on_parameter'
         # ./lib/rails_param/param_evaluator.rb:35:in `param!'
         # ./lib/rails_param/param.rb:3:in `param!'
         # ./spec/fixtures/controllers.rb:36:in `nested_array'
```

Reply:
`recurse_on_parameter` (or anything before) should raise exception if the parameter is provided and its value is invalid (i.e. `nil`), it shouldn't try to process it.

Reply:
@ifellinaholeonce this test is failing, because we've now changed the way we handle nils inside. My initial thought was just to go to recurse_on_parameter and simply add a `return nil if parameter.value.nil?` guard, but that will just make this test case fail, as it will no longer return an error and handle this the same way it handles the below cases. Thoughts?

Reply:
I have just applied that change again.

Reply:
@iMacTia but should it? If it should, then the next test should equally fail, because I am passing `filter: { state: nil }` explicitly in it, and that's what the current test is processing `filter: "state"` as.

Reply:
@iMacTia on the other hand, if we declare that `filter: "state"` is not valid then on what grounds? On the grounds that `"state"` is not a valid representation of a hash? If so #81 will fix that. If it's on the grounds that `"state"` is a valid representation of a hash, and it should process the hash as `{ state: nil }` and it is an array `state` that shouldn't be `nil` then the whole premise of this PR is reversed, because my whole point is that explicit `nil` should be valid.

Reply:
Yeah the two are equivalent, the only difference being that the first implicitly sets the array to nil, while on the latter that's explicit. They should both fail on the premise that the controller validates the array content

Reply:
@iMacTia okay, so the grounds for a failure is NOT the fact that the array here is `nil`, but rather the fact that the array here is `nil` when it's definition has the following line:

```
s.param! idx, String, required: true
```

Is that correct? The first two tests that we are talking about here should expect 400 for this:

```
      f.param! :state, Array do |s, idx|
        s.param! idx, String, required: true
      end
```

But should expect 200 for this:

```
      f.param! :state, Array do |s, idx|
        s.param! idx, String, required: true
      end
```

Reply:
Yep that's how I would expect it to work. I'm not saying that's how it works now, this probably need fixing, but that seems the logical thing to me.

I'm open to other thoughts as well though @ccdredzik @ifellinaholeonce

Reply:
Yeah that seems right to me. Thanks for spelling that out @ccdredzik.

Reply:
@iMacTia @ifellinaholeonce so I think this now works differently depending on whether `{ array: nil }` was passed explicitly, or was an attribute in a hash coerced from a string that had keys but not values. I will add a bunch of tests first to try to define all of the above cases so we can see what works and what doesn't.

Reply:
@iMacTia @ifellinaholeonce for the record: the more I dive into this, the more I think that the param block for array of anything else than a hash was a bad design choice. More sensible representation of the above would be:

```
param! :state, Array, of: String, allow_empty: false
```

I could understand a block if it would be an array of hashes:

```
param! :state, Array, of: Hash, allow_empty: false do |f|
  f.param! :a_param_of_a_hash, String
end
```

Reply:
Yeah we should support an array of hashes as well, that's pretty much the only meaningful scenario for the block. I also agree the shorthand you show above would make sense, the block seems a bit too much in that case

Reply:
@iMacTia and that block is also what makes it problematic to handle nils

Reply:
@iMacTia @ifellinaholeonce I have added an awful lot of tests. Can you double check if the assumptions they make are all correct?

Reply:
@iMacTia @ifellinaholeonce and a last thing for today: In param_spec.rb on line 472 there is a test case as follows:

```
it 'does not raise exception if array is not required but nested attributes are, and no array is provided' do
```

Which stands in direct contradiction to what we just talked about... I kind of want to give up.

Reply:
@ifellinaholeonce I've further discussed this with @ccdredzik offline, I agree there have been some confusion after the refactoring around this point. We agreed on dropping this test because, although the premises are good, it does not work well with hash and array coercion, unfortunately.

We all agree that a nil and missing params should be considered in the same way, so it would make sense to skip the recursive coercion when the parameter is provided, but is nil. Exactly as we do with a missing parameter.

This test can potentially be reintroduced later on, but we'll need to write it using a better payload system, as query-strings clearly don't work as expected and make it hard to reproduce the scenario we wanted to test

Reply:
I have pushed the implementation to the latest understanding of the contract. Please feel free to review.